ghproxy/ghcache/coalesce.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ghcache
+
+import (
+	"bufio"
+	"bytes"
+	"net/http"
+	"net/http/httputil"
+	"sync"
+)
+
+// requestCoalescer allows concurrent requests for the same key to share a
+// single upstream request and response.
+type requestCoalescer struct {
+	sync.Mutex
+	keys map[string]*responseWaiter
+}
+
+type responseWaiter struct {
+	*sync.Cond
+
+	resp *http.Response
+	err  error
+}
+
+// Request returns one of two things depending on if it there is already a
```

Comment:
I changed the requestCoalescer to implement the RoundTripper interface instead of trying to separate the synchronization from the request handling. It has a consistent interface and should be clearer now.